### ISSUE TITLE
AWS profile can't be found by kube-aws CLI

### ISSUE THREAD

**Issue by rguichard** [kubernetes-incubator/kube-aws#1231]
Hi,

I can't validate my cluster.yaml because kube-aws doesn't seem to find my aws profile...

```
$ kube-aws validate
Error: Failed to initialize cluster driver: failed to read/create encrypted assets: Error encrypting credentials/kubelet-tls-bootstrap-token: NoCredentialProviders: no valid providers in chain
caused by: EnvAccessKeyNotFound: AWS_ACCESS_KEY_ID or AWS_ACCESS_KEY not found in environment
SharedCredsLoad: failed to get profile
caused by: section 'customer-dev' does not exist
EC2RoleRequestError: no EC2 instance role found
caused by: RequestError: send request failed
caused by: Get http://169.254.169.254/latest/meta-data/iam/security-credentials: dial tcp 169.254.169.254:80: getsockopt: no route to host
```

My envs : 
```
$ env | grep -i aws
AWS_PROFILE=customer-dev
AWS_DEFAULT_PROFILE=customer-dev
AWS_CONFIG_FILE=/home/rguichard/.aws/config
AWS_DEFAULT_REGION=eu-west-1
AWS_HOME=/home/rguichard/.aws
```

and my /home/rguichard/.aws/config
```
[profile customer]
output = json
aws_access_key_id = [...]
aws_secret_access_key = [...]
region = eu-west-1
[profile customer-dev]
output = json
role_arn = arn:aws:iam::[...]:role/Development
source_profile = customer
region = eu-west-1
mfa_serial = arn:aws:iam::[...]:mfa/RGUICHARD_ext_Osones
```
This aws configuration seems to be ok because a simple `aws --profile customer-dev s3 ls` works perfectly.

What's wrong ? :/

**Comment by mumoshu** [comment]
@rguichard Hi. Thanks for trying kube-aws!

Hmm, unfortunately I have never used `.aws/config` but `.aws/credentials`.
Would there be any chance that the upstream aws-sdk-go which is vendored in kube-aws doesn't support the `.aws/config` yet but only `.aws/credentials`?

**Comment by rguichard** [comment]
Hello Mumoshu,

I tried to use `.aws/credentials` the same way I used `.aws/config` and the issue is the same.

**Comment by mumoshu** [comment]
@rguichard Thanks for the clarification!
Ah, then maybe `mfa_serial` isn't supported in the way we use aws-sdk-go?
What happens when you set `AWS_PROFILE=customer` - does the error differ?

**Comment by rguichard** [comment]
@mumoshu No it doesn't.

It seems that kube-aws needs to have an access key and a secret key in the profile and doesn't support to use a `source_profile`. I'll try to deploy my stack directly with CloudFormation without using kube-aws CLI.

**Comment by mumoshu** [comment]
@rguichard Thanks for the info!

It turns out to me that there's some parity between aws-sdk-go and aws-cli.
In nutshell, `source_profile` support is an opt-in feature.

So, does setting `AWS_SDK_LOAD_CONFIG` env var as [noted in the aws-sdk-go issue](https://github.com/aws/aws-sdk-go/issues/472#issuecomment-237342876) work for you?

**Comment by rguichard** [comment]
I did. But it seems the session option AssumeRoleTokenProvider must be set and as far I can see (Go is not my mother tongue...), it's not set in kube-aws ?

```
2018/04/13 11:04:51 ERROR: failed to create session with AWS_SDK_LOAD_CONFIG enabled. Use session.NewSession to handle errors occurring during session creation. Error: AssumeRoleTokenProviderNotSetError: assume role with MFA enabled, but AssumeRoleTokenProvider session option not set.
```

**Comment by mumoshu** [comment]
@rguichard Ah, good point! You might be correct. kube-aws [does miss that option in various places it creates AWS sessions](https://github.com/kubernetes-incubator/kube-aws/search?utf8=%E2%9C%93&q=%22session.New%22&type=).

Probably we need to rewrite every occurence of `session, err := session.NewSession(awsConfig)` to:

```golang
session, err := session.NewSessionWithOptions(session.Options{
	 Config: awsConfig,
         // This seems to be required for AWS_SDK_LOAD_CONFIG
	 SharedConfigState: session.SharedConfigEnable,
         // This seems to be required by MFA
         AssumeRoleTokenProvider: stscreds.StdinTokenProvider,
}
```

**Comment by mumoshu** [comment]
I'm working on this.

**Comment by mumoshu** [comment]
Submitted the fix at #1296
